### PR TITLE
Texture formats

### PR DESCRIPTION
**Pull request by kurtkuehnert**
# Objective

Currently some `TextureFormat`s are not supported by the Image type. 

## Solution
To remove that limitation I have removed the custom mapping from texture format to pixel size and replaced it with the wgpu equivalent.

### CODE REVIEW

File: crates/bevy_render/src/renderer/render_device.rs
Comment (superdump) [inline crates/bevy_render/src/renderer/render_device.rs:138]: I added this over here:
* https://github.com/bevyengine/bevy/pull/3884
* https://github.com/bevyengine/bevy/pull/3884/files#diff-a1595e1935032ef7580c1c4530e830f43efb3ba0e6c2de3aa2b3ffe788729091R130

Comment (superdump) [reply crates/bevy_render/src/renderer/render_device.rs:138]: That got merged so please update your PR on main.

Comment (Ku95) [reply crates/bevy_render/src/renderer/render_device.rs]: Done. :smile:

File: crates/bevy_render/src/texture/image.rs
Comment (superdump) [inline crates/bevy_render/src/texture/image.rs:419]: This is correct for non-block formats, but for block formats, it will be wrong. You kind of have to use both the block dimensions and the block byte size. It degrades to pixels as dimensions become 1x1.

**Review by superdump — CHANGES_REQUESTED** [review]
I wonder if we should just make use of wgpu's information directly and not bother wrapping it.

File: crates/bevy_render/src/texture/image.rs
Comment (Ku95) [reply crates/bevy_render/src/texture/image.rs:419]: Hmm I am not familiar with compressed textures / block formats, so I am not sure how to proceed. But my "soon to be" terrain plugin depends on being able to use some of the unlocked formats. 

Do you want to take over this PR?

Comment (Ku95) [reply crates/bevy_render/src/texture/image.rs:489]: Is this information even useful to have? I could just remove the pixel size thing entirely. 
It is used in some places in the engine though and I am not sure who else is using it.

Comment (Ku95) [reply crates/bevy_render/src/texture/image.rs]: @superdump I have updated the PR now. Is the block size guaranteed to be a multiple of its area? If not than this change is not correct.

Comment (superdump) [reply crates/bevy_render/src/texture/image.rs]: The block size is the size of a compressed block in bytes. So it accounts for all the data used to encode the block. I think I understand the intent of your question and the answer to that intent is yes. :)

**Review by superdump — COMMENTED** [review]
I think this looks good now, but I think it should be well tested before I approve it. :)

File: crates/bevy_render/src/texture/image.rs
Comment (Ku95) [reply crates/bevy_render/src/texture/image.rs]: Mhh, I think the current implementation of pixel size is wrong. 
For example the `TextureFormat::Astc5x4RgbaUnorm` has got a blocksize of 16 bytes and dimensions of 5 by 4 pixels. Thus the average pixel size is 6.4 bit/px. If that number is stored as a usize it is casted to 0, which is wrong. 
Prior to my PR only some of the uncompressed textures, where supported by the pixel size method, so this was not an issue.

This method is referenced in 14 places in the bevy code base and always used to calculate the size of the texture in bytes. 
By replacing `pixel_size` with `describe().block_size` we would enable bevys `Image` to be compatible with all uncompressed texture formats (This is the issue I am personally blocked on). 

However the same is not true for all compressed formats. 
I am not familiar with how to calculate their size properly. Do you have to divide the texture size in pixels by the block dimensions and subsequently multiply that by the block size? How would you deal with rounding? 
Does the texture size refer to the dimensions in pixels or in blocks? If the former: How to deal with sizes that are not a multiple of the block dimensions?

Comment (superdump) [reply crates/bevy_render/src/texture/image.rs]: Ah, I see. Good catch. Yes, I suppose maybe the real API then should match wgpu's in the sense that it should expose block dimensions and block size in bytes.

> However the same is not true for all compressed formats.
> I am not familiar with how to calculate their size properly. Do you have to divide the texture size in pixels by the block dimensions and subsequently multiply that by the block size? How would you deal with rounding?
Does the texture size refer to the dimensions in pixels or in blocks? If the former: How to deal with sizes that are not a multiple of the block dimensions?

Yes, pretty much. For calculating the size of a texture, you have to round up to the next block. So if you have a block size of 4x4 and a texture that is 6x6 then that will be 2 blocks wide and 2 blocks high. I think elsewhere I have calculated this as `image width in blocks = (image width in pixels + block width in pixels - 1) / block width in pixels` and similar for height. And then use the image dimensions in blocks to calculate the image size. Note then that for non-compressed formats, where the `block width` and `block height` are both 1, `image width in blocks` will be equivalent to `image width in pixels` and the `block size in bytes` would just be the pixel size in bytes.

Comment (Ku95) [reply crates/bevy_render/src/texture/image.rs]: I removed the flawed pixel size method in favor of the new texture size method. This should be calculating the texture size in bytes properly now.

**Comment by alice-i-cecile** [comment]
Fixed by #6820.

**Pull request closed without merging**
